kinto/schema_validation.py

```diff
@@ -0,0 +1,112 @@
+import colander
+from jsonschema import Draft4Validator, ValidationError, SchemaError, validate
+from pyramid.settings import asbool
+
+from kinto.core import utils
+from kinto.core.errors import raise_invalid
+from kinto.views import object_exists_or_404
+
+
+class JSONSchemaMapping(colander.SchemaNode):
+    def schema_type(self, **kw):
+        return colander.Mapping(unknown='preserve')
+
+    def deserialize(self, cstruct=colander.null):
+        # Start by deserializing a simple mapping.
+        validated = super().deserialize(cstruct)
+
+        # In case it is optional in parent schema.
+        if not validated or validated in (colander.null, colander.drop):
+            return validated
+        try:
+            check_schema(validated)
+        except ValidationError as e:
+            self.raise_invalid(e.message)
+        return validated
+
+
+def check_schema(data):
+    try:
+        Draft4Validator.check_schema(data)
+    except SchemaError as e:
+        message = e.path.pop() + e.message
+        raise ValidationError(message)
+
+
+def validate_schema(data, schema, ignore_fields=[]):
+    required_fields = [f for f in schema.get('required', []) if f not in ignore_fields]
+    # jsonschema doesn't accept 'required': [] yet.
+    # See https://github.com/Julian/jsonschema/issues/337.
+    # In the meantime, strip out 'required' if no other fields are required.
+    if required_fields:
+        schema = {**schema, 'required': required_fields}
+    else:
+        schema = {f: v for f, v in schema.items() if f != 'required'}
+
+    try:
+        validate(data, schema)
+    except ValidationError as e:
+        if e.path:
+            field = e.path[-1]
+        elif e.validator_value:
+            field = e.validator_value[-1]
+        else:
+            field = e.schema_path[-1]
+        e.field = field
+        raise e
+
+
+def validate_from_parent_schema_or_400(data, resource_name, request, ignore_fields=[]):
+    """Lookup in the parent objects if a schema was defined for this resource.
+
+    If the schema validation feature is enabled, if a schema is/are defined, and if the
+    data does not validate it/them, then it raises a 400 exception.
+    """
+    settings = request.registry.settings
+    schema_validation = 'experimental_collection_schema_validation'
+    # If disabled from settings, do nothing.
+    if not asbool(settings.get(schema_validation)):
+        return
+
+    # For records, there can be several schemas (collection level + bucket level).
+    schemas = []
+
+    bucket_id = request.matchdict["bucket_id"]
+    bucket_uri = utils.instance_uri(request, 'bucket', id=bucket_id)
+
+    # For records, the schema defined on the collection will be validated first.
+    if resource_name == "record":
```

Comment:
I'm a little nervous about this. It hard-codes part, but only part, of the containment hierarchy of Kinto. To be honest, I preferred that this logic be part of the Record resource.